src/worker/jobs/expiry_notification.rs

```diff
@@ -0,0 +1,188 @@
+
+impl BackgroundJob for CheckAboutToExpireToken {
+    const JOB_NAME: &'static str = "expiry_notification";
+
+    type Context = Arc<Environment>;
+
+    #[instrument(skip(env), err)]
+    async fn run(&self, env: Self::Context) -> anyhow::Result<()> {
+        let mut conn = env.connection_pool.get()?;
+        // Check if the token is about to expire
+        // If the token is about to expire, trigger a notification.
+        check(&env.emails, &mut conn)
+    }
+}
+// Check if the token is about to expire and send a notification if it is.
+fn check(emails: &Emails, conn: &mut PgConnection) -> anyhow::Result<()> {
+    info!("Checking if tokens are about to expire");
+    let expired_tokens =
+        crate::models::token::ApiToken::find_tokens_expiring_within_days(conn, EXPIRY_THRESHOLD)?;
+    // Batch send notifications in transactions.
+    const BATCH_SIZE: usize = 100;
+    for chunk in expired_tokens.chunks(BATCH_SIZE) {
+        conn.transaction(|conn| {
+            for token in chunk {
+                // Send notification.
+                let user = User::find(conn, token.user_id)?;
+                let Some(recipient) = user.email(conn)? else {
+                    return Err(anyhow!("No address found"));
+                };
+                let email = ExpiryNotificationEmail {
+                    name: &user.gh_login,
+                    token_name: &token.name,
+                };
+                match emails.send(&recipient, email) {
+                    Ok(_) => {
+                        // Update the token to prevent duplicate notifications.
+                        diesel::update(token)
+                            .set(
+                                crate::schema::api_tokens::expiry_notification_at
+                                    .eq(now.nullable()),
+                            )
+                            .execute(conn)?;
+                    }
+                    Err(e) => {
+                        error!("Failed to send email: {:?} to {}", e, recipient);
+                    }
+                }
+            }
+            Ok::<_, anyhow::Error>(())
+        })?;
+    }
+
+    Ok(())
+}
+
+#[derive(Debug, Clone)]
+struct ExpiryNotificationEmail<'a> {
+    name: &'a str,
+    token_name: &'a str,
+}
+
+impl<'a> Email for ExpiryNotificationEmail<'a> {
+    const SUBJECT: &'static str = "Your token is about to expire";
+
+    fn body(&self) -> String {
+        format!(
+            r#"Hi @{},
+
+We noticed your token "{}" will expire in about {} days.
+
+If this token is still needed, visit https://crates.io/settings/tokens/new to generate a new one.
+
+Thanks,
+The crates.io team"#,
+            self.name, self.token_name, EXPIRY_THRESHOLD
+        )
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::{
+        models::token::ApiToken, schema::api_tokens, test_util::test_db_connection,
+        typosquat::test_util::Faker, util::token::PlainToken,
```

Comment:
I'm not sure we should rely on typosquat-specific test utilities in other parts of the codebase, especially if they require us to change the existing test utilities.

since we're only using `faker.user()` it would probably be easier to use `NewUser::new()` directly and keep the existing test utility as-is, which would also reduce the amount of unrelated changes in the PR diff.

Reply:
Updated. Good point!